### DISCUSSION TITLE
Removing version number from Gradle dependencies

### DISCUSSION THREAD

**Discussion by sihutch** [openrewrite/rewrite#4485] [Q&A] [answered]
Hi All,

Does anyone know if there is an existing recipe that I can use to remove the version number from a grade dependency? We are moving to the spring dependency management plugin, so the versions will be centrally managed. I am aware that I can use find and replace but I would prefer to use a "gradle" recipe if one exists. 

Thanks in advance

Simon

**Comment by timtebeek** [comment]
I know @Jenson3210 has done some related work to standardize and clear out dependency versions. Perhaps he has some insights to share here?

**Reply by Jenson3210** [reply, marked as answer]
I would recommend staying away from Find and replace for this one. 
Problem is there are about 10+ ways in declaring dependencies/versions. 

Are there any variables in the input? 

Or is it only map and string notation with 'simple' strings? 
In that case, I think that it would not be too hard. 

I would go for a basic gradle recipe(RemoveVersionNotation), used in a spring recipe that migrates to dependency plugin(MigrateToDependencyPlugin)?

**Reply by sihutch** [reply]
Thanks both. @timtebeek In our case, the spring dependency management is provided by a custom plugin. Would you be interested in a pr that simply offers the removal of versions from gradle dependencies regardless of whether the resulting dependency can be resolved once the recipe has run? This would definitely be useful for us but I would be happy to submit a pr if you think it worthwhile.

**Reply by Jenson3210** [reply]
So you mean to say a recipe that receives the typical group/version names and strips the version of the string and map notation? 

It so happens that I am giving a live coding session on custom recipe development (internally) and am looking for a usecase. 

I was thinking of requesting the audience for good use cases that would deliver real value to them but could use this one if no other ideas are brought up. 

Of course, feel free to bring up your own PR if @timtebeek sees value. 

Personally, I do see value in the 2 recipes:

- Remove version declaration
- migrate to dep mgmt plugin:
  - remove versions from bom versions
  - add plugin

**Reply by sihutch** [reply]
@Jenson3210. I

"So you mean to say a recipe that receives the typical group/version names and strips the version of the string and map notation?" Yes, and optionally (defaults to false) removes a property name from gradle.properties if it is only referred to by the dependency being updated. 

I see value in those two recipes, too. We have a direct need for "Remove version declaration" but not for "migrate to dep mgmt plugin", so my employer may only let me work on the first initially. 

Sounds like a good use case for your session. Thanks for taking the time to reply.

**Reply by timtebeek** [reply]
Hi both! Great to follow the discussion here; In general I can say we try not to have recipes that could potentially be damaging, as it's hard to educate folks when exactly it's safe to make a particular change. As such stripping a dependency version is a risky operation; only when you're certain that you can fallback onto a managed dependency version would that be safe to apply.

We then prefer recipes that make a change, and adopt the idiomatic change that goes along with it. As a parallel: when changing a Maven parent we also check to introduce dependency management for dependencies no longer managed in the new parent, and strip now superfluous dependency versions where newly managed. I'd like such an approach here better: when the dependency management plugin is introduced or upgraded, strip away explicit versions that override a (now) managed version. No idea if that's feasible, but figured provided that context as input here.

**Reply by shanman190** [reply]
So if you're on Gradle 6 or newer, I'd just bypass the Spring Dependency Management plugin. The Gradle native replacement is Gradle Platforms (ex: `implementation(platform("org.springframework.boot:spring-boot-dependencies:<version>"))`).

With all of the information that we have stored in the `GradleProject` marker we should now be able to implement a similar recipe to Maven's `RemoveRedundantDependencyVersions`.

From an implementation standpoint, I'd probably scan for any platforms and resolve their Maven Pom. Then use the edit phase and lookup the managed dependencies against all platforms, if a dependency is managed and has the same or an older version, then remove the version.

For some more information directly from Spring: https://docs.spring.io/spring-boot/gradle-plugin/managing-dependencies.html

**Reply by shanman190** [reply]
Started a draft PR here: https://github.com/openrewrite/rewrite/pull/4492

**Reply by sihutch** [reply]
@shanman190 We use the Spring Plugin over Gradle platform as we are using the Spring Boot BOM and want to have "property-based customization of managed versions" I think that this is very common for Spring Boot / Gradle. Do you know if it would be possible to access the BOM set by the plugin in a similar way ?

**Reply by shanman190** [reply]
Possibly. 

Apologies in advance for the incoming wall of text...

But first, I want to throw out there that both platforms have the ability to change the versions in use. With the Spring Dependency Management plugin, they update the Maven properties and with Gradle Platforms they utilize additional constraints. Just to note, that there is already an existing recipe to update constraints within `rewrite-gradle` as well. From a usage perspective, the Spring property mechanism sometimes has advantages if the property is used across multiple GA coordinates within the BOM due to its simplicity of being updated for all coordinates.

Now back to your ask here and where the wrinkles start to come in...

If we look at the most basic Spring Boot Gradle project, you'll often see both plugins applied, but zero `dependencyManagement` configuration. This is because internal to the Spring Boot Gradle Plugin (ie. `org.springframework.boot`), it's reacting to the application of the Spring Dependency Management Plugin (ie. `io.spring.dependency-management`) and then adding the `org.springframework.boot:spring-boot-dependencies:<pluginVersion>` BOM to the list of imported Maven BOMs to use. From there, the Spring Dependency Management plugin is utilizing a Gradle Resolution Strategy to forcibly set all dependency versions to those found within the associated BOMs or overridden properties.

As part of my draft PR where I've been focusing on Gradle Platforms, so far I've been just looking at the source itself, but the more that I was thinking about it the more that I'm wondering if I shouldn't be looking for BOMs within a specific configuration. 

So to summarize some of the differences between the two dependency management implementations:

Spring Dependency Management Plugin:
* Uses resolution strategies to effect the current project's dependencies across all configurations
* Exposes a minimal way to change a Maven property in order to effect the resolved versions
* Utilizes _many_ `detachedConfiguration` configurations in order to resolve the dependency graph which results in a slower implementation
* Must specify BOMs per project
* Outside of the plugin being applied and the optional `dependencyManagement` extension, there is no way to identify that there has been a BOM applied.

Gradle Platforms:
* Native implementation
* Utilizes rich version constraints to further effect the dependency graph (bonus of being able to tell readers _why_ the constraint is there via the `because` clause)
* Rich versions information is persisted into the Gradle `module-metadata.json` which is then available transitively through consumers
* Able to apply different platforms and constraints per configuration, if desired.
* Constraints are applied during the resolution of resolvable configurations without the need for `detachedConfiguration` instances resulting in a faster implementation.
* Platforms are transitive, so can come from dependencies that are being included

So to be the most correct, we should be looking at the dependency configurations for Gradle platforms, but for the Spring Dependency Management plugin we would have to rely upon insider knowledge and the optional `dependencyManagement` extension block. Ultimately, this would result in two completely divergent implementations. So long story short, it's definitely possible. But as you can hopefully see now, I'm not sure how much it's worth it given the Gradle and Spring teams have also been having discussions on how to support a similar mechanism to their property-based overrides (not an easy, low-hanging fruit task).
